### PR TITLE
feat(presets): add config:best-practices preset

### LINKED ISSUE #16100: Create `config:best-practices` preset
### What would you like Renovate to be able to do?

We, as Renovate developers, want to have a way to tell our users: "here's what we recommend you do". A `config:best-practice` preset will help us do this.

Users can choose to extend from our `config:best-practice` preset. Nobody is "forced" into a workflow they may not like.

### If you have any ideas on how this should be implemented, please tell us here.

**Todo by Renovate programmer:**

- [x] Create new `config:best-practices` preset, which includes:
  - [x] Pin GitHub Actions to SHA
  - [x] Pin Docker digests
  - [x] Pin all `devDependencies`
  - [x] Any other important best-practice that the Renovate team recommends

**Todo by @HonkingGoose:**

- [x] Explain the new `config:best-practices` preset in the new best practices upgrade file in PR #22233
  - [x] Explain why we created the `config:best-practice(s)` preset
  - [x] Explain the reason behind each config option
  - [x] How to use our presets (put it in your `extends` array)
  - [x] Link to the preset config on the published docs

**Related issues:**

- #12024
- #11987

### Is this a feature you are interested in implementing yourself?

Partially, I'll let somebody else do the code work, and I'll do the documentation work.

### PR DESCRIPTION
**Pull request by DuncanCasteleyn**
## Changes

A new preset was added which allows you to configure what is considered best practice by the renovate team.

## Context

- Relates to #16100 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/internal/config.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/config.ts:19]:
```suggestion
    description: 'Preset with best practices from the Renovate maintainers.',
```

File: lib/config/presets/internal/config.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/config.ts:30]: Are there any more best practices that the Renovate maintainers want to add here?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some first comments.

File: lib/config/presets/internal/config.ts
Comment (DuncanCasteleyn) [reply lib/config/presets/internal/config.ts:18]: I like the idea, I'll wait for an opinion of another maintainer before applying the suggestion.

File: lib/config/presets/internal/config.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/config.ts:15]: Reminder: this PR targets `main`. We probably want to create a `v36` branch, and merge this PR into the `v36` branch. 😉

File: lib/config/presets/internal/config.ts
Comment (viceice) [reply lib/config/presets/internal/config.ts:18]: 🤷‍♂️

**Review by viceice — COMMENTED** [review]
🤷‍♂️

File: lib/config/presets/internal/config.ts
Comment (rarkins) [reply lib/config/presets/internal/config.ts:30]: We should check our own presets to see if any worth adding

Comment (HonkingGoose) [reply lib/config/presets/internal/config.ts:30]: I found some things in our [`default.json` in the `renovatebot/.github` repository](https://github.com/renovatebot/.github/blob/main/default.json) to think about.

### `docker:enableMajor`:

This one is a bit odd for me. I would expect extending from `config:base` would give you `major` Docker image upgrades out-of-the-box. There's probably a reason we don't enable `major` Docker images by default for `config:base`?

I think as a best practice, people should use the latest `major` version of their Docker images, or at least be aware that there's a new `major` version.

```json
{
  "packageRules": [
    {
      "enabled": true,
      "matchDatasources": [
        "docker"
      ],
      "matchUpdateTypes": [
        "major"
      ]
    }
  ]
}
```

### `group:linters`:

Grouping linters is usually safe, but I guess most people want separate updates?

```json
{
  "packageRules": [
    {
      "extends": "packages:linters",
      "groupName": "linters"
    }
  ]
}
```

And this is what `packages:linters` does:

```json
{
  "extends": [
    "packages:emberTemplateLint",
    "packages:eslint",
    "packages:stylelint",
    "packages:tslint"
  ],
  "matchPackageNames": [
    "prettier",
    "remark-lint",
    "standard"
  ]
}
```

### `configMigration=true`

After this issue is done:

- #16359

We should include `configMigration=true` in our best practices preset once it's ready to go. 😄

### Update version variables in shell scripts/GitHub Workflows:

I like the idea of this config, we'd need to check if this generalizes to other configs though:

```json
    {
      "description": "Update _VERSION variables in shell scripts",
      "fileMatch": ["\\.sh$"],
      "matchStrings": [
        "# renovate: datasource=(?<datasource>[a-z-]+?) depName=(?<depName>.+?)(?: (?:packageName|lookupName)=(?<packageName>.+?))?(?: versioning=(?<versioning>[a-z-]+?))?\\s[A-Z_]+?_VERSION=(?<currentValue>.+?)\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{versioning}}{{else}}semver{{/if}}"
    },
    {
      "description": "Update `version:` and `_VERSION:` variables in github workflows",
      "fileMatch": ["^\\.github/workflows/[^/]+\\.ya?ml$"],
      "matchStrings": [
        "\\s+version: (?<currentValue>.+?) # renovate: datasource=(?<datasource>.+?) depName=(?<depName>.+?)(?: (?:packageName|lookupName)=(?<packageName>.+?))?(?: versioning=(?<versioning>.+?))?\\s",
        "\\s*[A-Z_]+?_VERSION: (?<currentValue>.+?) # renovate: datasource=(?<datasource>.+?) depName=(?<depName>.+?)(?: (?:packageName|lookupName)=(?<packageName>.+?))?(?: versioning=(?<versioning>.+?))?\\s"
      ]
    }
```

### Group Prettier packages

This looks nice to have as well.

By the way, this looks like a good candidate for a new preset?

```json
    {
      "description": "Group prettier packages",
      "matchPackageNames": ["prettier"],
      "matchPackagePatterns": ["^@prettier/", "^prettier-plugin-"],
      "groupName": "prettier packages"
    },
```

### Dashboard Approval for `major` updates?

For me, I'd rather get the `major` updates automatically, but maybe you guys prefer the approval first flow?

```json
    {
      "description": "Require dashboard approval for major updates except Renovate",
      "excludePackagePatterns": [
        "^renovate(bot)?($|/)",
        "^(?:ghcr.io/)?containerbase/base$",
        "^@renovatebot/"
      ],
      "matchUpdateTypes": ["major"],
      "dependencyDashboardApproval": true
    },
```

Comment (DuncanCasteleyn) [reply lib/config/presets/internal/config.ts:30]: docker:enableMajor will probably have something to do about the fact that versions in docker images don't always follow semantic versioning standards?

Comment (DuncanCasteleyn) [reply lib/config/presets/internal/config.ts:30]: Is grouping something suited for a best practice preset?
Seems like something that should be decided by maintainers?
Is there a way to disable the grouping so users of the preset are not forced to copy paste the preset and remove the grouping to disable it?

if we add `configMigration=true` here then I assume it not added to `config:base`? if the plan is to add it to `config:base` then this would be inherited.

Comment (HonkingGoose) [reply lib/config/presets/internal/config.ts:30]: > `docker:enableMajor` will probably have something to do about the fact that versions in docker images don't always follow semantic versioning standards?

I don't know. 😄 


> Is grouping something suited for a best practice preset?
> Seems like something that should be decided by maintainers?
> Is there a way to disable the grouping so users of the preset are not forced to copy paste the preset and remove the grouping to disable it?

I'm not quite sure what you mean here. The `config:base` preset already groups certain packages and monorepos together. Do you mean that our new `config:best-practice(s)` preset should _not_ add any more groupings?

> if we add `configMigration=true` here then I assume it not added to `config:base`? if the plan is to add it to `config:base` then this would be inherited.

We have an issue that tracks setting config migration to `true` as the default behavior. I don't know _how_ we're going to make that the default behavior though: 🙈 

- https://github.com/renovatebot/renovate/issues/16359

File: lib/config/presets/internal/config.ts
Comment (marcusrbrown) [reply lib/config/presets/internal/config.ts]: `config:recommended`?

File: lib/config/presets/internal/config.ts
Comment (HonkingGoose) [reply lib/config/presets/internal/config.ts:30]: @rarkins can you review this thread, and let us know if there's anything you want to add to our opinionated `best-practices` preset? If you're happy with things as they are, let us know also. 😄

Comment (rarkins) [reply lib/config/presets/internal/config.ts:30]: - `docker:enableMajor` is no longer needed
- Skip `group:linters`
- Include `configMigration=true`
- We should probably convert our regex managers into presets but not sure there's any point including them by default
- I think dashboard approval for major updates is generally a good idea, but not certain whether to put it in config:recommended

Comment (rarkins) [reply lib/config/presets/internal/config.ts:30]: We should check if any of our org's regex managers are _not_ in these presets

Comment (HonkingGoose) [reply lib/config/presets/internal/config.ts:30]: > We should check if any of our org's regex managers are _not_ in these presets

I checked, but a maintainer will need to review my work. There are four repositories that have things that may be useful to others. See issue:

- #22025

File: lib/config/presets/internal/config.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/config.ts:18]: Merging this PR will close the linked issue. But that issue needs more work from @HonkingGoose after merging. Can you edit the top comment of this PR , like this:

```diff
+ Related issue #16100 
- closes https://github.com/renovatebot/renovate/issues/16100
```

After you've done that a maintainer probably needs to unlink this PR from the issue in the PR sidebar.

File: lib/config/presets/internal/config.ts
Comment (rarkins) [inline lib/config/presets/internal/config.ts:30]: Let's make this into a preset if one doesn't already exist (`:pinDevDependencies`)

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by HonkingGoose** [comment]
We already have two presets that pin dev dependencies. 

## `default:pinDevDependencies` [^docs]

> Pin dependency versions for `devDependencies`.
>
> ```json
> {
>   "packageRules": [
>     {
>       "matchDepTypes": [
>         "devDependencies"
>       ],
>       "rangeStrategy": "pin"
>     }
>   ]
> }
> ```

## `default:pinOnlyDevDependencies` [^pin-only]

> Pin dependency versions for `devDependencies `and retain SemVer ranges for others.
>
> ```json
> {
>   "packageRules": [
>     {
>       "matchPackagePatterns": [
>         "*"
>       ],
>       "rangeStrategy": "replace"
>     },
>     {
>       "matchDepTypes": [
>         "devDependencies"
>       ],
>       "rangeStrategy": "pin"
>     },
>     {
>       "matchDepTypes": [
>         "peerDependencies"
>       ],
>       "rangeStrategy": "widen"
>     }
>   ]
> }
> ```

[^docs]: https://docs.renovatebot.com/presets-default/#pindevdependencies
[^pin-only]: https://docs.renovatebot.com/presets-default/#pinonlydevdependencies

**Comment by rarkins** [comment]
Let's use the first

**Comment by rarkins** [comment]
Let's use the first

**Comment by rarkins** [comment]
@JamieMagee @viceice @secustor do you think "dashboard approval for major PRs" should be considered as best practice too? I know we always do it, but is it only possible because we are _diligent_?

**Comment by secustor** [comment]
I don't use it and mark the PRs instead with a `major` label. 
IMO hiding this behind approvals has the down side that users are not notified for updates ( e.g. Github notifications ) anymore.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by DuncanCasteleyn** [comment]
> @JamieMagee @viceice @secustor do you think "dashboard approval for major PRs" should be considered as best practice too? I know we always do it, but is it only possible because we are _diligent_?

https://github.com/renovatebot/renovate/pull/21239#issuecomment-1541356228 shows an example on how to do this, but it might be a good idea to add it as a preset?

**Comment by HonkingGoose** [comment]
> > @JamieMagee @viceice @secustor do you think "dashboard approval for major PRs" should be considered as best practice too? I know we always do it, but is it only possible because we are _diligent_?
> 
> https://docs.renovatebot.com/key-concepts/dashboard/#require-approval-for-major-updates shows an example on how to do this, but it might be a good idea to add it as a preset?

I like this idea, here my draft PR:

- #22075

**Comment by viceice** [comment]
why happens when the platform doesn't support dashboard?

**Comment by HonkingGoose** [comment]
> why happens when the platform doesn't support dashboard?

I think you mean _what happens_ here. And my answer is that I don't know what happens then. 😄

**Comment by DuncanCasteleyn** [comment]
> why happens when the platform doesn't support dashboard?

What happens when you have dependency dashboard is disabled and you enable dashboard approval is also something I'm wondering I don't see any checks on that.

**Comment by rarkins** [comment]
How about any of these we use in this repo?

```
  "semanticCommitScope": "deps",
  "automergeType": "pr",
  "prCreation": "immediate",
  "dockerfile": {
    "semanticCommitType": "build"
  },
  "packageRules": [
    {
      "matchPackageNames": ["containerbase/node"],
      "versioning": "node"
    },
    {
      "matchDepTypes": ["dependencies"],
      "semanticCommitType": "build"
    },
    {
      "matchPackageNames": ["semantic-release"],
      "semanticCommitType": "build"
    },
    {
      "matchPackageNames": ["@types/jest"],
      "groupName": "jest monorepo"
    },
    {
      "matchPaths": ["**/__fixtures__/**"],
      "enabled": false
    }
  ]
```

**Comment by rarkins** [comment]
Yes, I think it essentially ruins their workflow then, so we should clarify

**Comment by HonkingGoose** [comment]
I created a discussion for us to talk about the "Dependency Dashboard approval"-edgecase:

- #22298

**Comment by HonkingGoose** [comment]
@rarkins can you please read the comments in this PR and tell us what changes you want? This PR is getting stale because it needs maintainer-level input. 😉

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
I'm OK with it as-is

**Review by HonkingGoose — APPROVED** [review]
I'm okay with merging this preset. We can always expand the preset later if needed. 😄

File: lib/config/presets/internal/config.ts
Comment (mrazauskas) [inline lib/config/presets/internal/config.ts:22]: @viceice This should be `'config:recommended'`, or? I noticed that in documentation. Not sure what should be updated. Only this file and docs will be generate automatically?

Comment (HonkingGoose) [reply lib/config/presets/internal/config.ts:22]: The `config:best-practices` preset still uses `config:base` in the current `main` branch:

https://github.com/renovatebot/renovate/blob/c89eeb89dff3ee5a62b71c936bd447e579512e27/lib/config/presets/internal/config.ts#L1-L15

I think this was caused because we merged the `config:best-practices` preset to `main` before we merged the `config:base -> `config:recommended` PR. The `config:base` rename PR didn't fix up this file.

**Pull request merged**
